### PR TITLE
Fix products incorrectly marked as discounted

### LINKED ISSUE #10433: Cart block - Third party plugin compatibility - the price shows as "discounted" while it is not
## Describe the bug

When the cart item price is modified by a third party plugin and the price is different from the original, it shows as discounted, even though it is not. (see screenshot)

## To reproduce

1. Go to [https://demos.mklacroix.com/custom-laptop/](https://demos.mklacroix.com/custom-laptop/)
2. Select some options with an extra price
3. Add to cart
4. Visit the cart comparison page (https://demos.mklacroix.com/cart-block-vs-cart-shortcode/)

## Expected behavior

Either the original price should only be displayed if there is an actual discount applied, or there should be an option to not display it.

## Screenshots

image

### WordPress (please complete the following information):

-   WordPress version: 6.3
-   WooCommerce version: 7.9
-   WooCommerce Blocks version: 10.7.1 
-   Site language: EN
-   Any other plugins installed: [Product configurator for WooCommerce](https://wordpress.org/plugins/product-configurator-for-woocommerce/) and add-ons. The Extra price add-on changes the item price.

## Additional context

As the developper of the plugins mentioned above, this issue will be part of a series of issues brought forward in an effort to make them compatible with the cart and checkout blocks.

### PR DESCRIPTION
**Pull request by tarhi-saad**
## What

We are incorrectly displaying some products as discounted when their price is updated to add some extra cost.

Fixes #10433

## Why

Revised the `isDiscounted` logic. Now, the discount is applied only when the new price is lower than the `regularPrice`, rather than just being different.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. Add a discounted item and a regular-priced item to your cart.
2. Go to the Cart Block page. Confirm that the discount label is only visible next to the discounted product.
3. Go to the Checkout Block page. Under the `Order Summary` section, confirm that the discount label is only visible next to the discounted product.
4. Add the following code snippet using the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin:

```PHP
function calc_price( $cart_object ) {
    foreach ( $cart_object->get_cart() as $hash => $value ) {
        $value['data']->set_price( 50 );
    }
}
add_action( 'woocommerce_before_calculate_totals', 'calc_price' );
```

5. Go to the Cart Block page. Confirm that both products reflect the new price (50) and neither displays a discount label.
6. Go to the Checkout Block page, under the `Order summary` section, Confirm that both products reflect the new price (50) and neither displays a discount label.

* [ ] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## Screenshots or screencast

| Before | After |
| ------ | ----- |
|    image    |   image   |

## WooCommerce Visibility

Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [ ] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [x] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Fix products incorrectly marked as discounted

### CODE REVIEW

**Review by nielslange — COMMENTED** [review]
Thanks for working on this, @tarhi-saad. On the Cart and Checkout blocks, the change works as expected. However, I noticed that on the product page, the old price logic is still visible. 

Screenshot 2023-10-23 at 15 00 37<br>

While it's obvious to us, that @woocommerce/rubik maintains the Cart and Checkout blocks, while @woocommerce/woo-fse maintains all other blocks, for external folks, this might not be obvious. They might be wondering why this PR only partially changes the logic of how we display discounted prices.

I wonder if we should get someone from @woocommerce/woo-fse involved, to ensure that we change the display logic of all discounted prices. Or if we should create a follow-up ticket for them and ensure that both PRs will land in the same release.

cc: @ralucaStan, @tjcafferkey & @danielwrobert

**Comment by gigitux** [comment]
I think that it works as expected! The hook works on `before_calculate_totals`, so it makes sense that it impacts only the Cart/Checkout pages.

I tried with the storefront theme, and I noticed that the prices of the products don't change.

Futhermore, I noticed this weired behavior with the Mini Cart:

image

Instead, with the new Mini Cart block, this behavior has been fixed:
image

**Comment by gigitux** [comment]
Before merging this PR, I would suggest adding E2E tests to catch future potential regression. It should not require a lot of time and it could help us in the long term.

There is dedicated documentation that explains how testing filters and actions: https://github.com/woocommerce/woocommerce-blocks/blob/7403a5a4c2a23f1f5e94041e2ac1a7dde6dc8805/tests/e2e/mocks/custom-plugins/readme.md

**Comment by nielslange** [comment]
> I think that it works as expected! The hook works on before_calculate_totals, so it makes sense that it impacts only the Cart/Checkout pages.

Right. I haven't thought about that. Thanks for pointing this out, @gigitux.

**Review by nielslange — APPROVED** [review]
My previous comment was answered in https://github.com/woocommerce/woocommerce-blocks/pull/11386#issuecomment-1774851028. Let's ⛴️ this change, @tarhi-saad.

**Comment by ralucaStan** [comment]
Please 

> Before merging this PR, I would suggest adding E2E tests to catch future potential regression. It should not require a lot of time and it could help us in the long term.

+1 on this. @tarhi-saad don't merge this without adding a test. Let's strive to add automated tests to all our bug fixes.

**Review by tarunvijwani — COMMENTED** [review]
I'd be happy to pair and add E2E test cases with you! Let me know. 🙂

File: tests/e2e/tests/cart/cart-block.shopper.block_theme.spec.ts
Comment (tarhi-saad) [inline tests/e2e/tests/cart/cart-block.shopper.block_theme.spec.ts]: Hi, @gigitux! I followed the guidance from [Testing WordPress actions and filters](https://github.com/woocommerce/woocommerce-blocks/blob/7403a5a4c2a23f1f5e94041e2ac1a7dde6dc8805/tests/e2e/mocks/custom-plugins/readme.md#testing-wordpress-actions-and-filters) but didn't achieve the expected outcome. The custom plugin that's supposed to alter product prices in the Cart didn't seem to work; it showed the original prices instead. I tried debugging this issue yesterday without success. Could you help me in identifying any potential steps I might have overlooked?

Comment (gigitux) [reply tests/e2e/tests/cart/cart-block.shopper.block_theme.spec.ts]: Hi, @tarhi-saad!
I tried to run the test, and the price is updated.

image

I don't think that it is related, but the only change that I made was to add the suffix `.side_effects`. This is necessary because the filter could impact other tests (Mini Cart - for example), so it has to be run NOT in parallel.

**Comment by tarhi-saad** [comment]
All tests are passing! I'll merge this PR!

**Pull request merged**
